### PR TITLE
Sumo Logic Syslog Processor

### PR DESCRIPTION
**Pull request by sumo-drosiek**
Add Syslog Processor to keep backward compatibility with current solutions

# Sumo Logic Syslog Processor

Supported pipeline types: logs

The Sumo Logic Syslog processor can be used to create attribute with facility name
basing on facility code. Default facility name is syslog.

## Configuration

| Field         | Default  | Description                                                        |
|---------------|----------|--------------------------------------------------------------------|
| facility_attr | facility | The attribute name in which a facility name is going to be written |

## Examples

Following table shows example facility names which are extracted from log line

| log                       | facility            |
|---------------------------|---------------------|
| <13> Example log          | user-level messages |
| <334> Another example log | syslog              |
| Plain text log            | syslog              |

## Configuration Example

```yaml
processors:
  sumologic_syslog:
    facility_attr: testAttrName
```

### CODE REVIEW

**Review by pmm-sumo — APPROVED** [review]
This is nice processor, one thing I am wondering if perhaps a slightly different approach could have been taken and "transformprocessor" could be created. It would be more generic and would allow to transform attributes

File: processor/sumologicsyslogprocessor/testdata/sumologic_syslog_config.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+exporters:
+  nop:
+
+processors:
+  sumologic_syslog:
+    facility_attr: testAttrName
+
+service:
+  pipelines:
+    traces:
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/testdata/sumologic_syslog_config.yaml:13]:
```suggestion
    logs:
```

Wouldn't this make more sense ?

File: processor/sumologicsyslogprocessor/processor.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	}
+)
+
+func newSumologicSyslogProcessor(cfg *Config) *sumologicSyslogProcessor {
+	return &sumologicSyslogProcessor{
+		syslogFacilityAttrName: cfg.FacilityAttr,
+	}
+}
+
+func (ssp *sumologicSyslogProcessor) ProcessLogs(ctx context.Context, ld pdata.Logs) (pdata.Logs, error) {
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/processor.go:70]: Comment 🙏 ?

File: processor/sumologicsyslogprocessor/processor.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+					value string = syslogSource
+					ok    bool
+				)
+
+				log := logs.At(k)
+				r, _ := regexp.Compile(`^<(?P<number>\d+)>`)
+				match := r.FindStringSubmatch(log.Body().StringVal())
+
+				if match != nil {
+					facility, _ := strconv.Atoi(match[1])
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/processor.go:94]: Why wouldn't we check the error 🤷 ?
https://golang.org/pkg/strconv/#Atoi

Comment (sumo-drosiek) [reply processor/sumologicsyslogprocessor/processor.go:94]: being number is validated by regex

**Comment by sumo-drosiek** [comment]
@pmm-sumo We can iterate in that direction, but I don't have a general idea how this should work

**Comment by pmm-sumo** [comment]
> @pmm-sumo We can iterate in that direction, but I don't have a general idea how this should work

By moving the [mapping](https://github.com/SumoLogic/opentelemetry-collector-contrib/pull/1313/files#diff-4c8ea116ce21b6867dff96358bbfe02343f27fd581bfa47349f6c62b6c78a33cR36-R61) to config keys. E.g. 

```
 attrtibute: facility_id
 mapped_attribute: facility_name
 mapping:
   0: "kernel_messages"
   1: "user-level messages"
   ...
```

**Comment by sumo-drosiek** [comment]
This case is not so simple, because I need to extract value for mapping, divide it by 8 and this is my base for mapping.
Anyway, mapping processor is a good idea and it's possible that we will need it anyway :sweat_smile: :+1:

**Comment by pmm-sumo** [comment]
> I need to extract value for mapping, divide it by 8 and this is my base for mapping.

On the other hand, the mapping keys could be multiplied by 8, I guess? :)

**Comment by sumo-drosiek** [comment]
> On the other hand, the mapping keys could be multiplied by 8, I guess? :)

It could, but it would look like pretty ugly :cold_sweat:  

```
 mapping:
   0: "kernel_messages"
   1: "kernel_messages"
   2: "kernel_messages"
   3: "kernel_messages"
   4: "kernel_messages"
   5: "kernel_messages"
   6: "kernel_messages"
   7: "kernel_messages"
```

File: processor/sumologicsyslogprocessor/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Sumo Logic Syslog Processor
+
+Supported pipeline types: logs
+
+The Sumo Logic Syslog processor can be used to create attribute with facility name
+basing on facility code. Default facility name is syslog.
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/README.md:6]: nitpick:

```suggestion
based on facility code. Default facility name is `syslog`.
```

File: processor/sumologicsyslogprocessor/config.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sumologicsyslogprocessor
+
+import (
+	"go.opentelemetry.io/collector/config"
+)
+
+const ()
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/config.go:21]: ?

File: processor/sumologicsyslogprocessor/config.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+// Config holds the configuration for tail-based sampling.
+type Config struct {
+	*config.ProcessorSettings `mapstructure:"-"`
+
+	// FacilityAttr is attribute name when facility name should be placed
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/config.go:27]:
```suggestion
	// FacilityAttr is the name of the attribute the facility name should be placed into.
```

File: processor/sumologicsyslogprocessor/config_test.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+func TestLoadConfig(t *testing.T) {
+	factories, err := componenttest.NopFactories()
+	assert.NoError(t, err)
+
+	factory := NewFactory()
+	factories.Processors[factory.Type()] = factory
+
+	cfg, err := configtest.LoadConfigFile(t, path.Join(".", "testdata", "sumologic_syslog_config.yaml"), factories)
+	require.Nil(t, err)
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/config_test.go:36]:
```suggestion
	require.NoError(t, err)
```

File: processor/sumologicsyslogprocessor/config_test.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configtest"
+)
+
+func TestLoadConfig(t *testing.T) {
+	factories, err := componenttest.NopFactories()
+	assert.NoError(t, err)
```
Comment (pmalek-sumo) [inline processor/sumologicsyslogprocessor/config_test.go:30]:
```suggestion
	require.NoError(t, err)
```

File: processor/sumologicsyslogprocessor/config.go (hunk 5 above)
Comment (sumo-drosiek) [inline processor/sumologicsyslogprocessor/config.go:21]:
```suggestion

const ()
```

**Review by pmalek-sumo — APPROVED** [review]
🚀

**Pull request merged**